File: Strings/python/permutation_lexo.py
__author__ = "Shashwat Tiwari"
__email__  = "shashwat1791@gmail.com"

from collections import OrderedDict

class Permutation:

	def permute(self, text):
		_map = OrderedDict()
		for ch in text:	
			if ch in _map:
				_map[ch] += 1
			else:
				_map[ch] = 1
		
		string = []
		count = []
		
		for item in _map:
			string.append(item)
			count.append(_map[item])

		result = ['']*len(text)
		result_set = []

		self.permute_util(string, count, 0, result, result_set)
		return result_set

	def permute_util(self, string, count, level, result, result_set):
		# base case
		if level == len(result):
			result_set.append(''.join(result))
			return 

		for i in range(len(string)):
			if count[i] == 0:
				continue

			result[level] = string[i]
			count[i] -= 1
			self.permute_util(string, count, level+1, result, result_set)
			count[i] += 1


if __name__ == "__main__":
	p = Permutation()
	result = p.permute("AABC")

	for x in result:
		print x

File: Strings/PermutationLexo.java
/*
	@author  Shashwat Tiwari
	@email   shashwat1791@gmail.com
*/

import java.util.*;
import java.lang.*;

/*
	Given an input string, generate the permutations in lexographically sorted order

	Input:

	ABC
    
    Output:

    ABC
    ACB
    BAC
    BCA
    CAB
    CBA

*/
public class PermutationLexo{


	public List<String> permute(char [] input){
		// create a map, of character with counts
		TreeMap<Character, Integer> map = new TreeMap<>();
		for(char ch : input){
			map.compute(ch, (key, val) -> {
				if(val == null){
					return 1;
				}else{
					return val+1;
				}
			});
		}

		char [] str   = new char[map.size()];
		int  [] count = new int[map.size()];

		int index = 0;
		for(Map.Entry<Character, Integer> entry : map.entrySet()){
			str[index] = entry.getKey();
			count[index] = entry.getValue();
			index++;
		}


		List<String> resultSet = new ArrayList<>();
		char [] result = new char[input.length];

		permuteUtil(str, count, result, 0, resultSet);
		return resultSet;
	} 

	public void permuteUtil(char [] str, int [] count, char [] result, int level, List<String> resultSet){
		// base case
		if(level == result.length){
			resultSet.add(new String(result));
			return;
		}

		for(int i=0 ; i < str.length; i++){
			if(count[i] == 0){
				continue;
			}

			result[level] = str[i];
			count[i]--;
			permuteUtil(str, count, result, level + 1, resultSet);
			count[i]++;

		}

	}


	public static void main(String [] args){
		PermutationLexo test = new PermutationLexo();
		test.permute("AABC".toCharArray()).forEach(s -> System.out.println(s));
	}

}